btree/btree.c

```diff
@@ -2952,6 +2961,518 @@ static void ff_node_free_credits(const struct nd *node,
+ * +--------+-+--+----+--------+---------+---++-++-++-+--+--+--+--------------+----+--------+--+
+ *          | 0  5    12                 |    |  |  |                           ^       ^     ^
+ *          +----------------------------+    |  |  |                           |       |     |
+ *                                            +--+--+---------------------------+-------+-----+
+ *                                               |  |                           |       |
+ *                                               +--+---------------------------+-------+
+ *                                                  |                           |
+ *                                                  +---------------------------+
+ *
+ *
+ *
+ *  The above structure represents the way variable sized keys and values may be
+ *  stored in memory.
+ *  Node Hdr or Node Header will store all the relevant info regarding this node
+ *  type.
+ *  The keys wil be added fromstarting from the end of the node header in an
+ *  increasing order whereas the values will be added starting from the end of
+ *  the node such that the Value of the first record will be at the extreme
+ *  right, the Value for the second record to the left of the Value of the 
+ *  first record and so on.
+ *  This way the Keys start populating from the left of the node (after the 
+ *  Node Header) while the Values start populating from the right of the node.
+ *  As the records get added the empty space between the Keys list and the
+ *  Values list starts to shrink.
+ *
+ *  Additionally, we will maintain a directory in the central region of the node
+ *  which will contain the offsets to identify the address of the particular
+ *  key and value.
+ *  This Directory starts in the center of the empty space between Keys and
+ *  Values and will float in this region so that if addition of the new record
+ *  causes either Key or Value to overwrite the Directory then the Directory
+ *  will need to be shifted to make space for the incoming Key/Value; 
+ *  accordingly the Directory pointer in the Node Header will be updated. If no
+ *  space exists to add the new Record then an Overflow has happened and the new
+ *  record cannot be inserted in this node. The credit calculations for the
+ *  addition of the Record need to account for the moving of this Directory.
+ *
+ */
+#if 0
+struct dir_rec {
+	uint8_t key_offset;
+	uint8_t val_offset;
+};
+
+struct vkv_head {
+	struct m0_format_header  vkv_fmt;      /*< Node Header */
+	struct node_header       vkv_seg;      /*< Node type information */
+
+	/** 
+	 * The above 2 structures should always be together with node_header
+	 * following the m0_format_header.
+	 */
+
+	uint16_t                 vkv_used;     /*< Count of records */
+	uint8_t                  vkv_shift;    /*< Node size as pow-of-2 */
+	uint8_t                  vkv_level;    /*< Level in Btree */
+	uint32_t                 dir_offset;  /*< starting address of dir */
```

Comment:
Append vkv prefix.

Reply:
Updated.
Change in vkvv_init() for dir_offset is remaining and will be pushed in the next commit.